src/tests/query-builder.test.ts

```diff
@@ -0,0 +1,175 @@
+
+		expect(async () => {
+			await validateQuery(qb1.toQuery(), headers);
+		}).not.toThrow();
+	});
+
+	it(`test basic query chain`, async () => {
+		let q = new QueryBuilder(database).where("VendorID", "<=", 1);
+		let query = q.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "VendorID", operation: "<=", value: 1 },
+		]);
+
+		expect(async () => {
+			await validateQuery(query, headers);
+		}).not.toThrow();
+
+		q = q.orderBy("VendorID", "ASC");
+		query = q.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "VendorID", operation: "<=", value: 1 },
+		]);
+		expect(query.orderBy).not.toBeNull();
+		expect(query.orderBy).toEqual([{ key: "VendorID", direction: "ASC" }]);
+		expect(async () => {
+			await validateQuery(query, headers);
+		}).not.toThrow();
+
+		q = q.select(["VendorID", "store_and_fwd_flag", "fare_amount"]);
+		query = q.toQuery();
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "VendorID", operation: "<=", value: 1 },
+		]);
+		expect(query.orderBy).not.toBeNull();
+		expect(query.orderBy).toEqual([{ key: "VendorID", direction: "ASC" }]);
+		expect(query.select).not.toBeNull();
+		expect(query.select).toEqual([
+			"VendorID",
+			"store_and_fwd_flag",
+			"fare_amount",
+		]);
+	});
+
+	it(`test basic derived query chain`, async () => {
+		const q0 = new QueryBuilder(database).where("fare_amount", "==", 1);
+		let query = q0.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "fare_amount", operation: "==", value: 1 },
+		]);
+
+		let q1 = q0.orderBy("fare_amount", "DESC");
+		query = q1.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "fare_amount", operation: "==", value: 1 },
+		]);
+		expect(query.orderBy).not.toBeNull();
+		expect(query.orderBy).toEqual([{ key: "fare_amount", direction: "DESC" }]);
+
+		let q2 = q1.select(["fare_amount"]);
+		query = q2.toQuery();
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "fare_amount", operation: "==", value: 1 },
+		]);
+		expect(query.orderBy).not.toBeNull();
+		expect(query.orderBy).toEqual([{ key: "fare_amount", direction: "DESC" }]);
+		expect(query.select).not.toBeNull();
+		expect(query.select).toEqual(["fare_amount"]);
+	});
+
+	it(`test multi derived query chain`, async () => {
+		const q0 = new QueryBuilder(database).where("fare_amount", "==", 1);
+		let query = q0.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "fare_amount", operation: "==", value: 1 },
+		]);
+
+		let q1 = q0.where("VendorID", "==", 3);
+		query = q1.toQuery();
+
+		expect(query.where).not.toBeNull();
+		expect(query.where).toEqual([
+			{ key: "fare_amount", operation: "==", value: 1 },
+			{ key: "VendorID", operation: "==", value: 3 },
+		]);
+	});
+
+	it(`test green + red queries`, async () => {
```

Comment:
an important test case. Create a set of red and green queries from a derived state. Half of them should fail while the other half should pass. But they both originate from the same state. 

This test asserts the uniqueness of a query instance.